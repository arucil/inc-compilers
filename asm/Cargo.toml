[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
