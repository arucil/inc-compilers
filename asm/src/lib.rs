#![feature(never_type)]

use num_derive::{FromPrimitive, ToPrimitive};
use std::fmt::{self, Debug, Formatter, Write};

#[derive(Debug, Clone)]
pub struct Program<INFO = (), VAR = !> {
  pub info: INFO,
  pub blocks: Vec<(String, Block<VAR>)>,
}

#[derive(Clone)]
pub struct Block<VAR = !> {
  pub code: Vec<Instr<VAR>>,
}

#[derive(Clone)]
#[non_exhaustive]
pub enum Instr<VAR = !> {
  Add(Arg<VAR>, Arg<VAR>),
  Sub(Arg<VAR>, Arg<VAR>),
  Mov(Arg<VAR>, Arg<VAR>),
  Neg(Arg<VAR>),
  Call(String, usize),
  Ret,
  Push(Arg<VAR>),
  Pop(Arg<VAR>),
  Jmp(String),
  Syscall,
}

#[derive(Clone, PartialEq, Eq)]
pub enum Arg<VAR = !> {
  Imm(i64),
  Reg(Reg),
  Deref(Reg, i32),
  Var(VAR),
}

#[derive(
  Clone, Copy, FromPrimitive, ToPrimitive, PartialEq, Eq, PartialOrd, Ord, Hash,
)]
pub enum Reg {
  Rsp,
  Rbp,
  Rax,
  Rbx,
  Rcx,
  Rdx,
  Rsi,
  Rdi,
  R8,
  R9,
  R10,
  R11,
  R12,
  R13,
  R14,
  R15,
}

impl<INFO: Debug, VAR: Debug> Program<INFO, VAR> {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = format!("{:?}", self.info);
    for (label, block) in &self.blocks {
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }

  pub fn to_nasm(&self) -> String {
    let mut buf = format!(
      "extern read_int, print_int, print_newline\n\
      section .text\n"
    );
    for (label, block) in &self.blocks {
      buf += "\n";
      if label == "_start" {
        buf += "    global _start\n";
      }
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }
}

impl<VAR: Debug> Debug for Block<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    for instr in &self.code {
      write!(f, "    ")?;
      instr.fmt(f)?;
      writeln!(f)?;
    }
    Ok(())
  }
}

impl<VAR: Debug> Debug for Instr<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Add(src, dest) => write!(f, "add {:?}, {:?}", dest, src),
      Self::Mov(src, dest) => write!(f, "mov {:?}, {:?}", dest, src),
      Self::Call(label, _) => write!(f, "call {}", label),
      Self::Jmp(label) => write!(f, "jmp {}", label),
      Self::Neg(dest) => write!(f, "neg {:?}", dest),
      Self::Pop(dest) => write!(f, "pop {:?}", dest),
      Self::Push(src) => write!(f, "push {:?}", src),
      Self::Ret => write!(f, "ret"),
      Self::Sub(src, dest) => write!(f, "sub {:?}, {:?}", dest, src),
      Self::Syscall => write!(f, "syscall"),
    }
  }
}

impl<VAR: Debug> Debug for Arg<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Imm(n) => write!(f, "{}", n),
      Self::Var(var) => var.fmt(f),
      Self::Deref(r, i) => {
        if *i > 0 {
          write!(f, "qword [{:?} + {}]", r, i)
        } else if *i == 0 {
          write!(f, "qword [{:?}]", r)
        } else {
          write!(f, "qword [{:?} - {}]", r, -i)
        }
      }
      Self::Reg(r) => r.fmt(f),
    }
  }
}

impl Reg {
  pub fn is_callee_saved(&self) -> bool {
    match self {
      Reg::Rsp
      | Reg::Rbp
      | Reg::Rbx
      | Reg::R12
      | Reg::R13
      | Reg::R14
      | Reg::R15 => true,
      _ => false,
    }
  }
}

impl Debug for Reg {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    use Reg::*;
    let op = match self {
      Rsp => "rsp",
      Rbp => "rbp",
      Rax => "rax",
      Rbx => "rbx",
      Rcx => "rcx",
      Rdx => "rdx",
      Rsi => "rsi",
      Rdi => "rdi",
      R8 => "r8",
      R9 => "r9",
      R10 => "r10",
      R11 => "r11",
      R12 => "r12",
      R13 => "r13",
      R14 => "r14",
      R15 => "r15",
    };
    f.write_str(op)
  }
}
