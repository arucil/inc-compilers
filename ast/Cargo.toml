[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
