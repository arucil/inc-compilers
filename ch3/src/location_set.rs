use asm::{Arg, Reg};
use ast::IdxVar;
use indexmap::map::Entry;
use indexmap::IndexMap;
use num_traits::{FromPrimitive, ToPrimitive};
use smallvec::SmallVec;
use std::fmt::{self, Debug, Write};
use std::ops::Deref;

#[derive(Debug, Clone)]
pub struct LocationSet(SmallVec<[u32; 2]>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Location(usize);

#[derive(Debug, Clone)]
pub struct VarStore(IndexMap<IdxVar, Var>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Var(usize);

impl LocationSet {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }

  pub fn add_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if self.0.is_empty() {
      self.0.push(0);
    }
    self.0[i / 32] |= 1 << (i % 32);
  }

  pub fn remove_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if !self.0.is_empty() {
      self.0[i / 32] &= !(1 << (i % 32));
    }
  }

  pub fn add_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset >= self.0.len() {
      self.0.resize(offset + 1, 0);
    }
    self.0[offset] |= 1 << (i % 32);
  }

  pub fn remove_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset < self.0.len() {
      self.0[offset] &= !(1 << (i % 32));
    }
  }

  pub fn remove_caller_saved_regs(&mut self) {
    self.remove_reg(Reg::Rax);
    self.remove_reg(Reg::Rcx);
    self.remove_reg(Reg::Rdx);
    self.remove_reg(Reg::Rsi);
    self.remove_reg(Reg::Rdi);
    self.remove_reg(Reg::R8);
    self.remove_reg(Reg::R9);
    self.remove_reg(Reg::R10);
    self.remove_reg(Reg::R11);
  }

  pub fn add_argument_regs(&mut self, arity: usize) {
    const ARGUMENT_REGS: [Reg; 6] =
      [Reg::Rdi, Reg::Rsi, Reg::Rdx, Reg::Rcx, Reg::R8, Reg::R9];

    assert!(arity <= 6);
    for reg in &ARGUMENT_REGS[..arity] {
      self.add_reg(*reg);
    }
  }

  pub fn iter(&self) -> LocationIter {
    let current = if self.0.len() > 0 { self.0[0] } else { 0 };
    LocationIter {
      set: self,
      index: 0,
      current,
    }
  }

  pub fn write(&self, f: &mut impl Write, var_store: &VarStore) -> fmt::Result {
    write!(f, "{{")?;
    let mut comma = false;
    for elem in self {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{:?}", elem.to_arg(var_store))?;
    }
    write!(f, "}}")
  }
}

pub struct LocationIter<'a> {
  set: &'a LocationSet,
  index: usize,
  current: u32,
}

impl<'a> Iterator for LocationIter<'a> {
  type Item = Location;

  fn next(&mut self) -> Option<Location> {
    while self.current == 0 {
      self.index += 1;
      if self.index >= self.set.0.len() {
        return None;
      }
      self.current = self.set.0[self.index];
    }
    let i = self.current.trailing_zeros() as usize + self.index * 32;
    self.current &= self.current - 1;
    Some(Location(i))
  }
}

impl<'a> IntoIterator for &'a LocationSet {
  type Item = Location;
  type IntoIter = LocationIter<'a>;

  fn into_iter(self) -> Self::IntoIter {
    self.iter()
  }
}

fn reg_index(reg: Reg) -> usize {
  reg.to_usize().unwrap()
}

fn var_index(var: Var) -> usize {
  var.0 + 16
}

impl VarStore {
  pub fn new() -> Self {
    Self(IndexMap::new())
  }

  pub fn get(&mut self, var: IdxVar) -> Var {
    let len = self.0.len();
    match self.0.entry(var) {
      Entry::Occupied(var) => *var.get(),
      Entry::Vacant(entry) => *entry.insert(Var(len)),
    }
  }
}

impl Deref for VarStore {
  type Target = IndexMap<IdxVar, Var>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl Var {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    Arg::Var(var_store.get_index(self.0 - 16).unwrap().0.clone())
  }
}

impl Location {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    if self.0 < 16 {
      Arg::Reg(Reg::from_usize(self.0).unwrap())
    } else {
      Arg::Var(var_store.0.get_index(self.0 - 16).unwrap().0.clone())
    }
  }

  pub fn from_arg(arg: Arg<IdxVar>, var_store: &mut VarStore) -> Option<Self> {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => Some(reg.into()),
      Arg::Var(var) => Some(Self(var_store.get(var).0 + 16)),
      Arg::Imm(_) => None,
      Arg::ByteReg(_) => unimplemented!(),
    }
  }

  pub fn to_reg(&self) -> Option<Reg> {
    Reg::from_usize(self.0)
  }

  pub fn to_var(&self) -> Option<Var> {
    if self.0 >= 16 {
      Some(Var(self.0 - 16))
    } else {
      None
    }
  }
}

impl From<Reg> for Location {
  fn from(reg: Reg) -> Self {
    Self(reg.to_usize().unwrap())
  }
}
