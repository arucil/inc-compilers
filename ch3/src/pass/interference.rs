use super::liveness::Info as OldInfo;
use crate::location_set::{Location, LocationSet, VarStore};
use asm::{Block, Instr, Program, Reg};
use ast::IdxVar;
use indexmap::{IndexMap, IndexSet};
use petgraph::dot::{Config, Dot};
use petgraph::graph::{Graph, NodeIndex, UnGraph};
use std::fmt::{self, Debug, Formatter};

pub struct Info {
  pub locals: IndexSet<IdxVar>,
  pub conflicts: IndexMap<String, InterferenceGraph>,
  pub var_store: VarStore,
}

pub struct InterferenceGraph {
  pub graph: UnGraph<Location, ()>,
  pub nodes: IndexMap<Location, NodeIndex>,
}

pub fn build_interference(
  mut prog: Program<OldInfo, IdxVar>,
) -> Program<Info, IdxVar> {
  let var_store = &mut prog.info.var_store;
  let live = &prog.info.live;
  let conflicts = prog
    .blocks
    .iter()
    .map(|(name, block)| {
      (name.clone(), build_graph(block, &live[name], var_store))
    })
    .collect();

  Program {
    info: Info {
      locals: prog.info.locals,
      conflicts,
      var_store: prog.info.var_store,
    },
    blocks: prog.blocks,
  }
}

fn build_graph(
  block: &Block<IdxVar>,
  live: &[LocationSet],
  var_store: &mut VarStore,
) -> InterferenceGraph {
  let mut graph = InterferenceGraph {
    graph: Graph::new_undirected(),
    nodes: IndexMap::new(),
  };
  for (i, instr) in block.code.iter().enumerate() {
    add_instr_edges(instr, &live[i + 1], var_store, &mut graph);
  }
  graph
}

fn add_instr_edges(
  instr: &Instr<IdxVar>,
  after: &LocationSet,
  var_store: &mut VarStore,
  graph: &mut InterferenceGraph,
) {
  let mut add = |write_loc: Location| {
    let write_loc_node = graph.get_node(write_loc);
    for after_loc in after {
      if after_loc != write_loc {
        let after_loc_node = graph.get_node(after_loc);
        graph.add_edge(write_loc_node, after_loc_node);
      }
    }
  };

  match instr {
    Instr::Add(_, dest)
    | Instr::Sub(_, dest)
    | Instr::Neg(dest)
    | Instr::Pop(dest) => {
      if let Some(dest_loc) = Location::from_arg(dest.clone(), var_store) {
        add(dest_loc);
      }
    }
    Instr::Call(_, _) => {
      add(Reg::Rax.into());
      add(Reg::Rcx.into());
      add(Reg::Rdx.into());
      add(Reg::Rsi.into());
      add(Reg::Rdi.into());
      add(Reg::R8.into());
      add(Reg::R9.into());
      add(Reg::R10.into());
      add(Reg::R11.into());
    }
    Instr::Mov(src, dest) => {
      if let Some(dest_loc) = Location::from_arg(dest.clone(), var_store) {
        let src_loc = Location::from_arg(src.clone(), var_store);
        let dest_loc_node = graph.get_node(dest_loc);
        for after_loc in after {
          if after_loc != dest_loc && Some(after_loc) != src_loc {
            let after_loc_node = graph.get_node(after_loc);
            graph.add_edge(dest_loc_node, after_loc_node);
          }
        }
      }
    }
    Instr::Push(_) | Instr::Ret | Instr::Syscall | Instr::Jmp(_) => {}
    _ => {}
  }
}

impl InterferenceGraph {
  fn get_node(&mut self, loc: Location) -> NodeIndex {
    if let Some(&node) = self.nodes.get(&loc) {
      node
    } else {
      let node = self.graph.add_node(loc);
      self.nodes.insert(loc, node);
      node
    }
  }

  fn add_edge(&mut self, n1: NodeIndex, n2: NodeIndex) {
    if !self.graph.contains_edge(n1, n2) {
      self.graph.add_edge(n1, n2, ());
    }
  }
}

impl Debug for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    let graph = self.conflicts["start"]
      .graph
      .map(|_, var| var.to_arg(&self.var_store), |_, _| ());
    Dot::with_config(&graph, &[Config::EdgeNoLabel]).fmt(f)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use asm::{Arg, Block};
  use ch2::pass::instruction::Info as OldOldInfo;
  use insta::assert_snapshot;
  use maplit::hashmap;

  fn var(name: &str) -> Arg<IdxVar> {
    Arg::Var(IdxVar {
      name: name.to_owned(),
      index: 0,
    })
  }

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Mov(Imm(1), var("v")),
      Mov(Imm(42), var("w")),
      Mov(var("v"), var("x")),
      Add(Imm(7), var("x")),
      Mov(var("x"), var("y")),
      Mov(var("x"), var("z")),
      Add(var("w"), var("z")),
      Mov(var("y"), var("t")),
      Neg(var("t")),
      Mov(var("z"), Reg(Rax)),
      Add(var("t"), Reg(Rax)),
      Jmp("conclusion".to_owned()),
    ];
    let label_live = hashmap! {
      "conclusion".to_owned() => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![("start".to_owned(), Block { code })],
    };
    let prog = super::super::liveness::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Pop(Reg(Rdi)),
      Pop(Reg(Rsi)),
      Push(var("x")),
      Mov(Reg(Rbx), var("w")),
      Call("foo".to_owned(), 3),
      Add(Reg(Rax), var("w")),
      Jmp("conclusion".to_owned()),
    ];
    let label_live = hashmap! {
      "conclusion".to_owned() => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![("start".to_owned(), Block { code })],
    };
    let prog = super::super::liveness::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }
}
