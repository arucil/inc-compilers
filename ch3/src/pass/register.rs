use crate::location_set::{Location, Var};
use super::interference::{Info as OldInfo, InterferenceGraph};
use asm::{Program, Block, Reg};
use ast::IdxVar;
use indexmap::IndexSet;
use petgraph::graph::NodeIndex;
use std::collections::HashMap;
use std::cmp::Ordering;
use bittyset::BitSet;
use priority_queue::PriorityQueue;

pub struct Info {
  pub locals: IndexSet<IdxVar>,
  /// in bytes
  pub stack_space: usize,
}

pub fn allocate_registers(prog: Program<OldInfo, IdxVar>) -> Program<Info> {
  let conflicts = prog.info.conflicts;
  let blocks = prog.blocks.into_iter()
    .map(|(label, block)| {
      let conflicts = &conflicts[&label];
      (label, allocate_registers_block(block, conflicts))
    })
    .collect();
  Program {
    info: Info {
      locals: prog.info.locals,
      stack_space: 0,
    },
    blocks,
  }
}

fn allocate_registers_block(
  block: Block<IdxVar>,
  conflicts: &InterferenceGraph,
) -> Block {
  todo!()
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Color(i32);

#[derive(Debug, Clone, PartialEq, Eq)]
struct Vertex {
  location: Location,
  color: Option<Color>,
  saturation: BitSet,
}

impl PartialOrd for Vertex {
  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
    self.saturation.len().partial_cmp(&other.saturation.len())
  }
}

impl Ord for Vertex {
  fn cmp(&self, other: &Self) -> Ordering {
    self.saturation.len().cmp(&other.saturation.len())
  }
}

fn color_graph(
  graph: &InterferenceGraph,
  reg_colors: &HashMap<Reg, Color>,
) -> HashMap<Var, Color> {
  let mut queue = PriorityQueue::<NodeIndex, Vertex>::new();

  for node in graph.graph.node_weights() {
  }

  let graph = graph.graph.map(
    |_, location| {
      Vertex {
        location: *location,
        color: None,
        saturation: BitSet::new(),
      }
    },
    |_, _| (),
  );

  todo!()
}