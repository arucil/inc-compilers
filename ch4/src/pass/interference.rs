#[cfg(test)]
mod tests {
  use super::super::*;
  use asm::{Label, Program};
  use ast::IdxVar;
  use ch2::pass::instruction_selection::Info as OldOldInfo;
  use ch3::location_set::LocationSet;
  use ch3::pass::interference;
  use indexmap::indexset;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov v, i
      mov w, 42
      mov x, v
      add x, 7
      mov y, x
      mov z, x
      add z, w
      mov t, y
      neg t
      mov rax, z
      add rax, t
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("v"),
          IdxVar::new("i"),
          IdxVar::new("w"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call foo, 3
      add w, rax
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn mov_same_variables() {
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov t, x
      add t, y
      mov z, t
      add z, w
    "#,
    );
    let label_live = hashmap! {};
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("t"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("w"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn if_in_init() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov tmp.0, rax
    cmp tmp.0, 3
    jge block3
    jmp block4
block0:
    mov rax, 2
    jmp conclusion
block1:
    mov rax, 41
    jmp conclusion
block2:
    cmp x.0, 10
    sete al
    movzx tmp.1, al
    cmp tmp.1, 0
    je block1
    jmp block0
block3:
    mov x.0, 10
    jmp block2
block4:
    mov x.0, 77
    jmp block2
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("tmp.0"),
          IdxVar::new("x.0"),
          IdxVar::new("tmp.1"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn nested_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        call read_int
        mov x, rax
        call read_int
        mov y, rax
        cmp x, 1
        jl block2
        jmp block3
      block0:
        mov rax, y
        add rax, 2
        jmp conclusion
      block1:
        mov rax, z
        add rax, 10
        jmp conclusion
      block2:
        cmp x, 0
        je block0
        jmp block1
      block3:
        cmp x, 2
        je block0
        jmp block1
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn complex_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov x.0, rax
    cmp x.0, 100
    jg block8
    jmp block9
block0:
    mov rax, -1
    jmp conclusion
block1:
    mov rax, y.1
    neg rax
    jmp conclusion
block2:
    mov tmp.1, y.1
    sub tmp.1, x.0
    cmp tmp.1, 10
    jl block0
    jmp block1
block3:
    cmp x.0, y.1
    jl block2
    jmp block1
block4:
    mov tmp.0, x.0
    sub tmp.0, y.1
    cmp tmp.0, 10
    jl block0
    jmp block3
block5:
    mov rax, 5000
    jmp conclusion
block6:
    mov rax, x.0
    jmp conclusion
block7:
    cmp x.0, 60
    jl block5
    jmp block6
block8:
    call read_int
    mov y.1, rax
    cmp x.0, y.1
    jge block4
    jmp block3
block9:
    cmp x.0, 40
    jg block7
    jmp block6
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x.0"),
          IdxVar::new("y.1"),
          IdxVar::new("tmp.1"),
          IdxVar::new("tmp.0"),
        },
      },
      constants: Default::default(),
      blocks,
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }
}
