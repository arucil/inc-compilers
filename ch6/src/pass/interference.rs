#[cfg(test)]
mod tests {
  use asm::{Label, Program};
  use ast::{IdxVar, Type};
  use ch3::location_set::LocationSet;
  use ch4::pass::instruction_selection::Info as OldOldInfo;
  use ch4::pass::interference;
  use ch4::pass::liveness_analysis;
  use indexmap::indexmap;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn vector() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Tuple(vec![Type::Int]),
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }

  #[test]
  fn str() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Str,
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }
}
