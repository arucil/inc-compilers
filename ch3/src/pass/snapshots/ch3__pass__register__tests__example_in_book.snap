---
source: ch3/src/pass/register.rs
expression: result.to_string_pretty()

---
locals: {}
used_callee_saved_regs: {}
stack_space: 16 bytes
start:
    mov qword [rbp - 8], 1
    mov rcx, 42
    mov qword [rbp - 8], qword [rbp - 8]
    add qword [rbp - 8], 7
    mov qword [rbp - 16], qword [rbp - 8]
    mov qword [rbp - 8], qword [rbp - 8]
    add qword [rbp - 8], rcx
    mov rcx, qword [rbp - 16]
    neg rcx
    mov rax, qword [rbp - 8]
    add rax, rcx
    jmp conclusion

